GitUI/GitUIExtensions.cs

```diff
@@ -34,82 +34,90 @@ public static class GitUIExtensions
-            {
-                return string.Empty;
-            }
+            openWithDiffTool ??= OpenWithDiffTool;
 
-            if (file.IsSubmodule)
+            if (file.IsNew || firstId == null || FileHelper.IsImage(file.Name))
             {
-                return LocalizationHelpers.ProcessSubmodulePatch(diffViewer.Module, file.Name, patch);
+                // View blob guid from revision, or file for worktree
+                return fileViewer.ViewGitItemRevisionAsync(file, selectedRev.ObjectId, openWithDiffTool);
             }
 
-            return patch.Text;
-        }
+            string selectedPatch = GetSelectedPatch(fileViewer, firstId, selectedRev.ObjectId, file);
+            return fileViewer.ViewPatchAsync(file.Name, text: selectedPatch ?? defaultText,
+                openWithDifftool: openWithDiffTool, isText: file.IsSubmodule || selectedPatch == null);
 
-        public static Task ViewChangesAsync(this FileViewer diffViewer,
-            [CanBeNull] ObjectId firstId,
-            [CanBeNull] GitRevision selectedRevision,
-            [NotNull] GitItemStatus file,
-            [NotNull] string defaultText,
-            [CanBeNull] Action openWithDifftool = null)
-        {
-            if (firstId == null && selectedRevision != null)
+            void OpenWithDiffTool()
             {
-                firstId = selectedRevision.FirstParentGuid;
+                fileViewer.Module.OpenWithDifftool(
+                    file.Name,
+                    file.OldName,
+                    firstId?.ToString(),
+                    selectedRev?.ToString(),
+                    "",
+                    file.IsTracked);
             }
 
-            openWithDifftool ??= OpenWithDifftool;
-            if (file.IsNew || firstId == null || FileHelper.IsImage(file.Name))
+            static string GetSelectedPatch(
+                FileViewer fileViewer,
+                ObjectId firstId,
+                ObjectId selectedId,
+                GitItemStatus file)
             {
-                // The previous commit does not exist, nothing to compare with
-                if (selectedRevision == null)
+                if (firstId == ObjectId.CombinedDiffId)
                 {
-                    throw new ArgumentNullException(nameof(selectedRevision));
+                    var diffOfConflict = fileViewer.Module.GetCombinedDiffContent(selectedId, file.Name,
+                        fileViewer.GetExtraDiffArguments(), fileViewer.Encoding);
+
+                    return string.IsNullOrWhiteSpace(diffOfConflict)
+                        ? Strings.UninterestingDiffOmitted
+                        : diffOfConflict;
                 }
 
-                // View blob guid from revision, or file for worktree
-                return diffViewer.ViewGitItemRevisionAsync(file, selectedRevision.ObjectId, openWithDifftool);
-            }
+                if (file.IsSubmodule && file.GetSubmoduleStatusAsync() != null)
```

Comment:
`file.GetSubmoduleStatusAsync() != null` this looks sus. An async method should be awaited, not checked for `null`s...

Reply:
This pattern is used in many situations (this PR reduces the usage, some more with #7825).
Submodule status is not always evaluated, this checks if the async Task has been started, if so await the result on the next line. (This is the normal code path I believe.)
If not started async, the submodule status is evaluated below.